src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "opto/addnode.hpp"
+#include "opto/convertnode.hpp"
+#include "opto/vmaskloop.hpp"
+
+//        L O O P   V E C T O R   M A S K   T R A N S F O R M A T I O N
+// ============================================================================
+
+// -------------------------------- Constructor -------------------------------
+VectorMaskedLoop::VectorMaskedLoop(PhaseIdealLoop* phase) :
+  _phase(phase),
+  _igvn(&(phase->_igvn)),
+  _arena(phase->C->comp_arena()),
+
+  _lpt(nullptr),
+  _cl(nullptr),
+  _cle(nullptr),
+  _iv(nullptr),
+
+  _core_set(_arena),
+  _body_set(_arena),
+  _body_nodes(_arena, 32, 0, nullptr),
+  _rpo_idx(_arena, 32, 0, 0),
+  _elem_bt(_arena, 32, 0, T_ILLEGAL),
+  _stmts(_arena, 2, 0, nullptr),
+  _swptrs(_arena, 8, 0, nullptr),
+  _size_stats(_arena)
+{}
+
+// ------------------- Entry function of vector masked loop -------------------
+void VectorMaskedLoop::try_vectorize_loop(IdealLoopTree* lpt) {
+  assert(UseMaskedLoop, "Option should be enabled");
+  assert(lpt->is_counted(), "Loop must be counted");
+  assert(lpt->is_innermost(), "Loop must be innermost");
+
+  CountedLoopNode *cl = lpt->_head->as_CountedLoop();
+  // Skip if loop is already vector masked
+  if (cl->is_vector_masked()) return;
+  // Skip non-post loop
+  if (!cl->is_post_loop()) return;
+  // Skip malformed counted loop
+  if (!cl->is_valid_counted_loop(T_INT)) return;
+  // Skip loop if stride is unsupported
+  if (abs(cl->stride_con()) != 1) return;
+  // Skip loop with control flow
+  if (cl->loopexit()->in(0) != cl) return;
+  // Skip if some loop operations are pinned to the backedge
+  if (cl->back_control()->outcnt() != 1) return;
+
+  // Init data structures and collect loop nodes
+  init(lpt);
+  if (!collect_loop_nodes()) return;
+
+  // Collect loop statements and analyze vectorizability
+  if (!collect_statements()) return;
+  if (!analyze_vectorizability()) return;
+
+  // Try creating a vector mask with the smallest vector element size
+  const TypeVectMask* t_vmask = create_vector_mask_type();
+  if (t_vmask == nullptr || !t_vmask->isa_vectmask()) return;
+
+  // Transform the loop and set flags
+  transform_loop(t_vmask);
+  cl->mark_loop_vectorized();
+  cl->mark_vector_masked();
+  _phase->C->set_max_vector_size(MaxVectorSize);
```

Comment:
On AArch64 with SVE, we pre-initialize an all-true register (p7) only in compiled methods where `MaxVectorSize` is set. So if a compiled method implicitly uses ptrue (because it's vectorized), we need to set the `MaxVectorSize` to guarantee p7 is initialized. This usage starts since the initial SVE patch (in 2020). We know this is not a perfect solution and @fg1417 is currently investigating if we have better solutions.